### PR TITLE
Chunk blobs into window size to avoid window overrun

### LINKED ISSUE #447: Assertion in windowing code about null window pos
Seeing this assertion in a leader + client setup. Cuda enabled, erasure not enabled:
```thread 'solana-broadcaster' panicked at 'assertion failed: win[pos].is_none()', src/streamer.rs:483:13```

### PR DESCRIPTION
**Pull request by sakridge**
Fixes #447

### CODE REVIEW

File: src/streamer.rs [hunk 1]
```diff
@@ -445,63 +445,70 @@ fn broadcast(
     while let Ok(mut nq) = r.try_recv() {
         dq.append(&mut nq);
     }
-    let mut blobs: Vec<_> = dq.into_iter().collect();
+    let blobs_vec: Vec<_> = dq.into_iter().collect();
+    let blobs_chunked: Vec<_> = blobs_vec
+        .chunks(window.read().unwrap().len())
```
Comment (garious) [inline src/streamer.rs]: That's fancy enough for a very long code comment.

File: src/streamer.rs [hunk 2]
```diff
@@ -445,63 +445,70 @@ fn broadcast(
...
-        ).is_err()
+        // Fill in the coding blob data from the window data blobs
+        #[cfg(feature = "erasure")]
         {
-            return Err(Error::GenericError);
+            if erasure::generate_coding(
+                &mut window.write().unwrap(),
+                *receive_index as usize,
+                blobs_len,
+            ).is_err()
```
Comment (garious) [inline src/streamer.rs]: Looks like that could be reduced to something like:
`.map_err(|_| Error::GenericError)?`

File: src/streamer.rs [hunk 3]
```diff
@@ -445,63 +445,70 @@ fn broadcast(
...
+            for b in &blobs {
+                let ix = b.read().unwrap().get_index().expect("blob index");
+                let pos = (ix as usize) % WINDOW_SIZE;
+                if let Some(x) = &win[pos] {
+                    trace!(
+                        "popped {} at {}",
+                        x.read().unwrap().get_index().unwrap(),
+                        pos
+                    );
+                    recycler.recycle(x.clone());
```
Comment (garious) [inline src/streamer.rs]: Strange place for a `clone()`. Is that necessary?

Comment (sakridge) [reply src/streamer.rs]: from @aeyakovenko yes, it's necessary. It's an element in the window array, so cannot be moved out..?

Comment (sakridge) [reply src/streamer.rs]:
```
warning: build failed, waiting for other jobs to finish...
error[E0507]: cannot move out of borrowed content
   --> src/streamer.rs:484:38
    |
484 |                     recycler.recycle(*x);
    |                                      ^^ cannot move out of borrowed content
```

File: src/streamer.rs (hunk 2 above)
Comment (sakridge) [reply src/streamer.rs]: +1

File: src/streamer.rs (hunk 3 above)
Comment (garious) [reply src/streamer.rs]: #451

**Comment by garious** [comment]
Merge in #451

**Pull request closed without merging**